signal.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2022 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPSignalE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"fmt"
+	"os"
+	"os/signal"
+	"sync"
+)
+
+// ShutdownSignal is a signal that caused the application to exit.
+type ShutdownSignal struct {
+	Signal os.Signal
+}
+
+// String will render a Signal type as a string suitable for printing.
+func (sig ShutdownSignal) String() string {
+	return fmt.Sprintf("%v", sig.Signal)
+}
+
+type signalReceivers struct {
+	m    sync.Mutex
+	last *ShutdownSignal
+	done []chan os.Signal
+}
+
+func (recv *signalReceivers) Done() chan os.Signal {
```

Comment:
minor: we should return a read only channel

```suggestion
func (recv *signalReceivers) Done() <-chan os.Signal {
```